text/0000-explode.md

```diff
@@ -0,0 +1,236 @@
+
+Right now, every addon in the ecosystem is allowed to implicitly depend on any feature in Ember. You can't know *a priori* which features are safe to remove if you're using any given addon.
+
+Therefore, we propose new metadata in the `ember-addon` section of `package.json`:
+
+```js
+  "ember-addon": {
+    "required-core-features": "3.1.0",
+    "required-additional-features": ["@ember/routing", "@ember/prototype-etxensions", "@ember/string"]
+  }
+```
+
+`required-core-features` records the latest version of `@ember/core` that the addon has been tested against. It causes your addon to depend on all the Ember packages that were still bundled inside that `@ember/core` version. 
+
+`required-additional-features` lets the addon express dependency on features that have already been split from `@ember/core` as of the base features version.
+
+
+### Why required-core-features instead of peerDependencies?
+
+`required-core-features` says something different than a `peerDependency` constraint.
+
+We want addons to be as aggressive as possible in bumping their `required-core-features`, because that creates the maximum opportunity for apps to pick their packages *a la carte*.
+
+Simultaneously, we want addons to support the widest practical range of Ember versions (which would be their `peerDependencies` constraint on Ember, if they choose to have one), because that makes everybody's upgrades easier. 
+
+For example, consider an addon with:
+
+```js
+  "ember-addon": {
+    "required-core-features": "3.1.0",
+    "required-additional-features": ["@ember/routing"]
+  }
+```
+
+If you use this addon in an app with `ember-source` 2.18.0, there's no problems, the metadata isn't affecting anything.
+
+If you use this addon in an app with `@ember/core` 3.6.0, there's still no problems, but we will be able to warn the user that they need to install `@ember/routing` plus whatever future packages happen to get split out of `@ember/core` between 3.1 and 3.6.
+
+In contrast, if there was a peerDependency on `@ember/core` we would break usage with `ember-source`, and *vice versa*. And if there was a peerDependency on `@ember/routing` it would break usage with any Ember version earlier than the one that split out that package.
+
+The "pure NPM" way of solving this problem is to release new versions of your addon to support new versions of Ember, but this creates a much more fragmented ecosystem than we want. We are careful not to break things very often, such that many addons really don't need to be released just to keep working with newer Ember releases.
+
+
+## Local Dependency Overrides
+
+We have already proposed making missing dependencies a hard build error. This is important so that weird and frustrating bugs don't crop up unexpectedly. But it needs to be possible to get oneself unstuck without being blocked by third-party code.
+
+Therefore, we propose a `config/dependency-overrides.js` file that allows you to make declarations on behalf of any addon:
```

Comment:
Ember should be striving for zero configuration. This adds lots more configuration that will need to be regularly updated if a user wants to use a recent version of Ember. Right now, on Ember Observer, I did a quick search and found just 17 addons whose `package.json` is up to date with the most recent version of Ember.

I strongly encourage you to rethink this.

Reply:
None of this configuration is required in any app. This is purely about making it *not impossible* to spend more effort if you want to make your app smaller.

Apps are free to ignore 100% of this RFC and will *always* be free to do so.

Reply:
The question for me is to what extent this RFC will impact addon developers, not app developers.